### PR TITLE
Added GameMapEx CopyTileMap function

### PR DESCRIPTION
**Pull request by TechCor8**

### CODE REVIEW

**Comment by TechCor8** [comment]
From Daniel Stevens on previous Pull Request:
> I'm afraid a `memcpy` wouldn't work here. Particularly not for arbitrary alignments. The internal format of map data is not a simple 2D array. It's more like a 2D array sliced up into a 3D array. The end result likely has better cache friendliness for most common operations, though makes it more involved to copy a section out to a normal 2D array.
> 
> I guess a part of me is wondering why we would want to copy to a normal 2D array. This seems rather use case specific. It would seem more in the spirit of these library to access the data in place.

There is a performance issue related to requesting each tile individually when requesting every tile. Since HFL is the primary access point for this type of operation, it makes sense to me to put it there. The question would be how else could a high performance copy be done, if not through HFL. Perhaps, the raw tilemap object is accessible outside of HFL so that the copy could be performed there. I will check.

**Comment by DanRStevens** [comment]
I'm suggesting the copy should be eliminated. You should be able to read the original in-place just as quickly as you can read a copy. Is the copy really required?

If you do indeed need a copy, you can potentially do a `memcpy`, but only if you preserve the same format. That would be easy for a full map copy. It wouldn't work so well for a partial map copy, as the parameters here suggest is being done.

File: Source/GameMapEx.cpp [hunk 1]
```diff
@@ -114,3 +114,29 @@ int GameMapEx::LoadMap(char *fileName)
...
+		{
+			int px = x - xMin;
+			int py = y - yMin;
+
+			int xLower = p->tileXMask & x;
+			int xUpper = xLower >> 5;
+			xLower &= 31;
+			xUpper <<= p->logTileHeight;
+			xUpper += y;
+			xUpper <<= 5;
```
Comment (DanRStevens) [inline Source/GameMapEx.cpp]: As per my earlier comment, this section of code could be extracted into an `inline` function. It converts the 2D `(x, y)` coordinates into a 1D array index. I believe OP2Utility has such a method.

**Review by DanRStevens — COMMENTED** [review]
So yeah, same comment as before, and of course I kind of question why a copy is being made. Particularly to a format that is not native to OP2.

A copy would make more sense to me if it copied from one map to another map, or from one section of the map to another section of the same map. That doesn't look like what is being done here. It looks like the map data is simply being copied to a standard 2D array.

For reference, OP2 breaks up the `x` coordinate into 2 components, and then indexes into what is essentially a 3D array: `tileData[xUpper][y][xLower]`

**Comment by TechCor8** [comment]
For context, the copy is being used to perform an asynchronous/threaded read for complex AI calculations in C#. There are a couple reasons why performing a direct read is bad in this case:
- OP2 will be modifying the tile map during threaded calculations which expect data to remain unchanged for a period of time.
- P/Invoke is a fairly slow operation, which is compounded when performing a massive number of requests. This means a copy in managed memory is essential.

**Comment by DanRStevens** [comment]
Is there a reason for the bound boxes? And have you considered using the same array format from C#?

**Comment by TechCor8** [comment]
I perform [tilePosition.x + tilePosition.y * m_Width] in C# because it's easy for me to understand. This other format is arcane to me. The C# implementation could be something different since it would be wrapped in the accessor method: GetCellType(LOCATION tilePosition), but again, I have no idea what the upper/lower/mask code is doing. It's copied from somewhere else. The conversion is to make it something I know how to use.

By bounding box, you must be talking about the min/max parameters? It is passing the playable area which is something like (32,0,160,127).

**Comment by TechCor8** [comment]
Are you hinting that I should copy the source array verbatim, and use the fancy upper/lower/mask code in C#? I think I had considered that, but could not figure out the size of the original tile array. Normally, I would assume width*height, but I didn't trust that assumption here.

**Comment by DanRStevens** [comment]
The upper and lower mask code can be written another way:
```cpp
xLower = x % 32;
xUpper = x / 32;
```

I suppose non-around-the-world maps will have the tile data offset by one column of 32. Other than that, you can `memcpy` from there to get the full array data. The size of the tile data is `width * height` elements, and they are contiguous in memory. They are re-arranged from normal order though.

Example array ordering for 64x4 map:
00..31    128..159
32..63    160..191
64..95    192..223
96..127  224..255

Notice the maps are sliced into columns, where each column is 32 tiles wide. After completing one column, it continues on to the next.

----

Edit: If you just want the full map data, I would probably suggest using `memcpy`. The bounding box dimensions wouldn't be needed for that. You will need to know the map width and height though, both for the size of the copy, and to determine the offset to the start of the data. The offset is either 0 for around-the-world-maps, or `32*height` for non-around-the-world maps. Around-the-world is determine by map `width`, with all values of 512+ being around-the-world.

**Comment by TechCor8** [comment]
I've been trying to get this to work without much luck.

In GameMapEx::CopyTileMap, I changed the for loop to memcpy:
```
int **tileArray = (int**)(mapTileData);
memcpy(tileMap, *tileArray, GetMapWidth()*GetMapHeight());
```

This bubbles up to C# where it is stored in a `uint[]`.

GameMapEx does not expose `tileXMask` and `logTileHeight`, so I can't access that externally where I need to convert to the tile index.

In C#, I've tried this:
```
tilePosition = GetPointInGridSpace(tilePosition); // Subtracts bounds min: (65,36) becomes (33,36)
int xLower = tilePosition.x & 0x1F; // ... 0001 1111
int xUpper = tilePosition.x >> 5;   // ... 1110 0000
int tileIndex = (xUpper * m_Height + tilePosition.y) * 32 + xLower; // m_Height = 128
// (65,36) = 5249 tileIndex
```

which is what OP2Utility is doing, and I also tried the `% /` technique above. Seems like I'm getting back junk data.

I might actually have to try and understand the map format...

**Comment by TechCor8** [comment]
This is my understanding of the example array shown above:

```
const int columnWidth = 32;
int column = x / columnWidth;
int columnStartIndex = column * (mapHeight * columnWidth);
int index = columnStartIndex + (x % columnWidth) + (y * columnWidth);
```

Sound about right?

I'm curious about what the edge cases are. For example, on6_01.map starts on the second column... Why? Do all non-ATW start on the second column, and all ATW maps start on the first one?

I'm going to test the above out soon.

--
I assume this weirdness is some kind of 1997 optimization technique. I'm confused by the bit masks, shifting, and logHeight use. Looks like some wacky attempt at avoiding divisions, which I guess were very slow ops back in the day. Not worth the readability hit these days.

Edit:
Something I noticed is GameMapWidth() is returning 256 for a 128 map.

Edit 2:
Okay, I think I figured out this insanity. The 128x128 map is actually 256x128, but everything is offset by 32 tiles (1 column), with a ton of dead space at the end. This has the side effect of having to copy twice as much map than necessary. Makes me wonder if an actual 256 map is 256, since 512 will make it ATW.

**Comment by TechCor8** [comment]
DotNetMissionSDK is now using the memcpy version of the CopyTiles method.

Something weird is going on. Looks like CopyTiles was already merged. I had to resolve a conflict from main.

If this change is acceptable, I'll let you guys merge it.

**Comment by DanRStevens** [comment]
The math looks right.

Around the world maps allocate exactly as much space as needed. Non-around-the-world maps allocate double, and then offset the data. I assume that's so there is padding data on either side of the map. I'm guessing there are some poorly handled boundary conditions, and that was a hack to get things to work.

Around-the-world is determined by the user seen width, not the internal padding width.

**Comment by Brett208** [comment]
I suspect when I merged the UnitEx change, it brought in this code as well for some reason, which is playing some havoc on the other PRs.

It would be nice to add some comment in the header on the use case for this function. It sounds like something along the lines of creating a copy of the game data to allow threaded calculations to occur without worrying about state change on the main game loop.

Otherwise, I'm fine with the code existing in HFL. This seems like an okay addition to the library. I would lean on @DanRStevens if he has any more implementation concerns.

File: Source/GameMapEx.cpp [hunk 2]
```diff
@@ -115,28 +115,9 @@ int GameMapEx::LoadMap(char *fileName)
...
-			xLower &= 31;
-			xUpper <<= p->logTileHeight;
-			xUpper += y;
-			xUpper <<= 5;
-			
-			tileMap[px + py * mapWidth] = (*tileArray)[xUpper + xLower];
-		}
-	}
+	
+	memcpy(tileMap, *tileArray, GetMapWidth()*GetMapHeight()*sizeof(int));
```
Comment (DanRStevens) [inline Source/GameMapEx.cpp]: The second parameter should be offset for non-around-the-world maps. As the copy size is bounded to the valid data size, you'll want to skip over the padding to ensure you return all of the correct data.

In the following, the compiler will automatically handle the `* sizeof(TileData)` as part of the pointer offset arithmetic. C++ is weird that way, with how addition to pointers is auto scaled by the element size.
```cpp
std::size_t offset = GetMapWidth() >= 512 ? 0 : 32 * GetMapHeight();
memcpy(tileMap, *tileArray + offset, GetMapWidth()*GetMapHeight()*sizeof(int));
```

Comment (TechCor8) [reply Source/GameMapEx.cpp]: GetMapWidth() is returning the internal double width. For what you are describing, it would have to be divided by two?
Actually, skipping the padding is a good optimization.

Comment (DanRStevens) [reply Source/GameMapEx.cpp]: You would be correct then.

I kind of feel like that function should not be returning the internal padding width though. It would seem to make more sense if it returned the user visible width. Maybe we should discuss that in a separate issue.

Comment (TechCor8) [reply Source/GameMapEx.cpp]: I'm going to check clipRect to see if I can detect the visible area so I can remove padding. Removing padding will affect how this function is used.

Comment (DanRStevens) [reply Source/GameMapEx.cpp]: If padding is removed by the copy, the only relevant attributes are then `width` and `height`. In a way, this simplifies how the result is used.

**Comment by TechCor8** [comment]
Copy now only includes visible area.
Tested on newworld.map and on6_01.map.

File: Source/GameMapEx.cpp [hunk 3]
```diff
@@ -117,7 +117,28 @@ int GameMapEx::LoadMap(char *fileName)
...
+		// Around the world map
+		padding = 0;
+		length = p->tileWidth * p->tileHeight;
+	}
+	else
+	{
+		// Standard map
+		padding = p->clipRect.x1 * p->tileHeight;
+		length = p->clipRect.Width() * p->clipRect.Height();
+	}
```
Comment (DanRStevens) [inline Source/GameMapEx.cpp]: I would be hesitant to involve the `clipRect` in this. That's a secondary value, which just happens to coincide with map dimensions. There is no guarantee the `clipRect` will relate to map padding. A map editor could potentially modify `clipRect` independently, should they wish to provide some kind of custom effect.

I would recommend going directly from map dimensions, which are directly tied to padding. I gave some formulas earlier in the thread. They should look something like:
```cpp
std::size_t paddingOffset = mapWidthInTiles >= 512 ? 0 : 32 * mapHeightInTiles;
std::size_t dataSize = mapWidthInTiles * mapHeightInTiles * sizeof(TileData); // sizeof(TileData) == 4
```

Unless perhaps you've encountered something that says otherwise?

----

One additional thought: I don't think `clipRect` relates to the visible portion, but rather the portion over which movement is allowed. Notably, I think the bottom row of the map tends to be excluded from the `clipRect`. In addition, a number of maps mark the bottom row as non passable.

Comment (TechCor8) [reply Source/GameMapEx.cpp]: That's good to know about clipRect. I may need to double check some C# SDK code as it is making some assumptions based on that being the visible map area.

The issue is that I can't find the visible map size. `tileWidth` returns the double size. I've noticed that `paddingOffsetTileX` returns 32, which may help avoid the magic number you are proposing. Still need the visible size, though. Maybe I can always divide by 2 if `paddingOffsetTileX` is 32?

Comment (TechCor8) [reply Source/GameMapEx.cpp]: Here's my new proposed code:
```
int mapWidth = p->paddingOffsetTileX == 32 ? p->tileWidth / 2 : p->tileWidth;
memcpy(tileMap, *tileArray + p->paddingOffsetTileX*p->tileHeight, mapWidth*p->tileHeight*sizeof(int));
```

Comment (TechCor8) [reply Source/GameMapEx.cpp]: Rereading your comment, I'd like to point out that the bottom row is not visible on "on6_01.map". Bottom right coordinates reported from the game's status bar shows 128,127. This is also true on ATW maps: 512,255

So the bottom row of tiles is always not visible on all maps.

Comment (Brett208) [reply Source/GameMapEx.cpp]: @TechCor8, would you mind adding a small comment before the function declaration explaining the use case?

I think for the use case of AI calculations, it probably doesn't need the bottom row since it has to be impassable. Maybe there could be some lava or blight on that row, but otherwise, there probably isn't anything to take into account.

Comment (DanRStevens) [reply Source/GameMapEx.cpp]: Huh, you're right. The map editors do show an extra row of tiles at the bottom of maps, which doesn't appear in game. I used to think that was just a difference in coordinate offsets affecting display of coordinates.

I may need to double check some of these data structures against the original code. It's been a while since I've looked at it, so the details may be getting fuzzy. I noticed the data structure in HFL has both `tileWidth`, `tileHeight`, and `logTileWidth`, `logTileHeight`. The methods `GetMapWidth()` and `GetMapHeight()` use the log versions.

Comment (TechCor8) [reply Source/GameMapEx.cpp]: I was browsing OP2Utility map.h:
```
/**
* \brief	Represents the visible areas of the map.
*
* \note	Maps designated 'around the world' allow for continuous
*			scrolling on the X axis and so will populate X1 with -1
*			and X2 with \c INT_MAX.
*/
Rect clipRect;
```
Comment seems to indicate that the clipRect is the visible area.

Comment (TechCor8) [reply Source/GameMapEx.cpp]: Did we decide what to do with this?
We can use clipRect or paddingOffsetTileX.

The current code could be slimmed down to:
```
if (p->clipRect.x1 == -1)
    memcpy(tileMap, *tileArray, p->tileWidth*p->tileHeight*sizeof(int));
else
    memcpy(tileMap, *tileArray + p->clipRect.x1*p->tileHeight, p->clipRect.Width()*p->clipRect.Height()*sizeof(int));
```
or use the paddingOffsetTileX and divide for the width:
```
int mapWidth = p->paddingOffsetTileX == 32 ? p->tileWidth / 2 : p->tileWidth;
memcpy(tileMap, *tileArray + p->paddingOffsetTileX*p->tileHeight, mapWidth*p->tileHeight*sizeof(int));
```

Comment (DanRStevens) [reply Source/GameMapEx.cpp]: You know, maybe we should look at merging this, and just opening an issue to revisit any areas we are uncertain about. It seems like the code is far enough along, and sounds like it's basically working.

----

I think I prefer having the single call to `memcpy`, with arguments determined up front.

----

I'm looking at some of the map loading code right now. The tile data loading is checking `Map.paddingOffsetTileX` (offset: `0x38`) to determine if the tile data should be loaded to an offset address. I'll look more into the rest of the details shortly, assuming I don't get called away.

Comment (DanRStevens) [reply Source/GameMapEx.cpp]: Ok, yes, looks like `paddingOffsetTileX` is the value to use to determine padding size, which can replace the hardcoded `32` (if you want). I noticed the value is set when there is padding, but left unset when there is no padding. I assume it is initialized elsewhere to 0.

----

The check to determine if the map is around the world, is when `lgMapTileWidth >= 9` (tileWidth >= 512). If that is not the case (`lgMapTileWidth < 9`), then we have a regular map. For a regular map, the `paddingOffsetTileX` field is set to 32 (hardcoded), and the `lgMapTileWidth` value is incremented by 1. That means if `lgMapTileWidth == 9`, it's unclear if that was the original value, or the incremented value. The adjusted value is stored at offset `0x14`.

When map data is loaded in from the file, it checks `paddingOffsetTileX` to determine the offset to load to. If that value is zero, it simply loads it to the array, without any offset or size adjustment. The size used is `Map.numTiles` (offset: `0x2C`) multiplied by `sizeof(TileData)` (`4`).

If `paddingOffsetTileX` is non-zero, then it calculates `&tileData[paddingOffsetTileX / 32, 0, paddingOffsetTileX % 32]`, and writes the data contiguously to that location. Note that if `paddingOffsetTileX % 32 != 0`, this would behave in a rather unexpected manner, however, due to the internal uses of base 2 log values, a minimum map size of 32 tiles width would prevent that condition from ever being true. For the size, it decrements the `Map.lgTileWidth` field (offset: `0x14`), and uses that to calculate map size. The calculation is: `Map.tileHeight << (Map.lgTileWidth - 1) * sizeof(TileData)`, where `sizeof(TileData) == 4`.

----

Hope that helps.

**Comment by Brett208** [comment]
I'm okay if it is merged. I just wanted to see some sort of high level comment on how and when to appropriately use CopyTileMap.

**Comment by TechCor8** [comment]
Replaced `clipRect` with `paddingOffsetTileX` and other width/height fields, and added some documentation per request.

FYI: Normal use in C++ will be pretty difficult as not enough data is exposed in GameMapEx. For example, it is impossible to know the real width externally. `GetMapWidth` returns double for standard maps, which can be 512, yet 512 can also be an ATW map. This is confirmed in Dan's comment above. I suggest a possible second set of methods such as `GetVisibleWidth` and `GetVisibleHeight`.

Second, calculating the index is a bit overwhelming. Although, once you have the visible width, it is possible to do. I still think a `GetTileIndex` method would be useful in this case. Of course, if no one needs CopyTileMap, it may be pointless to add.

--
Unrelated to this PR, but I seem to be getting an Access Violation crash when playing a 128x128 map and then following with an ATW map, but no other combination. Don't know how long it's been going on. It's not CopyTileMap (disabled it to check). Maybe I can narrow it down at some point. Disassembler points me to `77971D9C  ret         0Ch`.

**Review by DanRStevens — APPROVED** [review]

**Comment by DanRStevens** [comment]
Yeah, I'm kind of thinking `GetWidth()` should return the normal width, not the padded width. I'd prefer that being the default, and perhaps offer a `GetPaddedWidth()` if you wanted the internal size. Not sure how existing code uses the method though.

We should provide a `GetTileIndex` method. Perhaps a new PR for that.

For the crash, maybe send details as to what maps and how you loaded them. I checked the address, but it's not in Outpost2.exe's address space (or any loaded module).

**Comment by TechCor8** [comment]
I figured out the crash. Totally on my C# SDK. Since JSON missions share the same C# DLL and DLLs are never unloaded, some static objects were holding data about the unloaded mission. In this case, the old map size was smaller than the new one, so it was accessing out of bounds, triggering an access violation in ntdll.dll (wow64).

**Comment by DanRStevens** [comment]
Ahh, ok, glad you got that sorted out.
